### PR TITLE
Add C SmoothL1 criterion

### PR DESCRIPTION
**Pull request by szagoruyko**
with a doc by @andreaskoepf from https://github.com/torch/nn/pull/374
CUDA PR is coming

### CODE REVIEW

File: generic/SmoothL1Criterion.c [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+static int nn_(SmoothL1Criterion_updateOutput)(lua_State *L)
+{
+  THTensor *input = luaT_checkudata(L, 2, torch_Tensor);
+  THTensor *target = luaT_checkudata(L, 3, torch_Tensor);
+  int sizeAverage = luaT_getfieldcheckboolean(L, 1, "sizeAverage");
+  real sum;
+
+  sum = 0;
+  TH_TENSOR_APPLY2(real, input, real, target,
+                   real z = fabs(*input_data - *target_data);
```
Comment (soumith) [inline generic/SmoothL1Criterion.c]: fabs will work correctly for double tensors. For float, you have to use fabsf.
These are the C99 prototypes for the function:
``C
     double fabs  (double x);
      float fabsf (float x);
``

Comment (szagoruyko) [inline generic/SmoothL1Criterion.c]: what would be the difference with https://github.com/torch/nn/blob/master/generic/AbsCriterion.c#L14 ?

Comment (soumith) [inline generic/SmoothL1Criterion.c]: ugh, i suppose that is equally incorrect.

**Comment by andreaskoepf** [comment]
Isn't torch itself using fabs for FloatTensors as well? see https://github.com/torch/torch7/blob/086ec2a760a15a5165b488e25899f3ee70e2ec6e/TensorMath.lua#L946

File: generic/SmoothL1Criterion.c (hunk 1 above)
Comment (andreaskoepf) [inline generic/SmoothL1Criterion.c]: Isn't torch itself using fabs for FloatTensors as well? see https://github.com/torch/torch7/blob/086ec2a760a15a5165b488e25899f3ee70e2ec6e/TensorMath.lua#L946

Comment (fmassa) [inline generic/SmoothL1Criterion.c]: yes, @andreaskoepf is right, `torch` has different `abs` for `int` and `long`, but for `double` and `float` it uses `fabs`
For a reference, look at [fabs](https://github.com/torch/torch7/blob/master/lib/TH/generic/THTensorMath.c#L1844), [labs](https://github.com/torch/torch7/blob/master/lib/TH/generic/THTensorMath.c#L1801) and [abs](https://github.com/torch/torch7/blob/master/lib/TH/generic/THTensorMath.c#L1805)

Comment (soumith) [inline generic/SmoothL1Criterion.c]: sergey said he'll correct this disrepancy in a separate PR overall for torch.

**Pull request merged**